### DISCUSSION TITLE
Deploying a transparent upgradeable proxy contract and calling the initializer function

### DISCUSSION THREAD

**Discussion by damien75** [zkSync-Community-Hub/zksync-developers#25] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

0.8.16

### zksync-web3 Version

0.14.3

### Hardhat.config.ts

{
  "name": "greeter-example",
  "version": "1.0.0",
  "main": "index.js",
  "author": "...",
  "license": "MIT",
  "devDependencies": {
    "@matterlabs/hardhat-zksync-deploy": "^0.6.3",
    "@matterlabs/hardhat-zksync-solc": "^0.3.17",
    "@matterlabs/hardhat-zksync-upgradable": "^0.0.1-alpha.3",
    "@matterlabs/hardhat-zksync-verify": "^0.1.7",
    "@matterlabs/signature-checker": "^1.1.1",
    "@nomiclabs/hardhat-etherscan": "^3.1.7",
    "@openzeppelin/contracts": "^4.9.0",
    "@openzeppelin/contracts-upgradeable": "^4.9.0",
    "@openzeppelin/hardhat-upgrades": "^1.27.0",
    "ethers": "^5.7.2",
    "hardhat": "^2.14.1",
    "ts-node": "^10.9.1",
    "typescript": "^5.1.3",
    "zksync-web3": "^0.14.3"
  },
  "dependencies": {
    "dotenv": "^16.1.3",
    "solc": "^0.8.20"
  }
}

### Deployment Script (WITHOUT PRIVATE KEY)

import { Wallet } from "zksync-web3";
import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";
import * as hreUpgrades from "hardhat";
import { getInitializerData } from "@matterlabs/hardhat-zksync-upgradable/dist/src/utils/utils-general";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
import { config } from "dotenv";
config();

// An example of a deploy script that will deploy and call a simple contract.
export default async function (hre: HardhatRuntimeEnvironment) {
  console.log(`Running deploy script for the Greeter contract`);

  // Get the private key from dotenv.
  const privateKey = process.env.WALLET_PRIVATE_KEY;
  if (!privateKey) {
    throw new Error("Private key not found in environment variables");
  }
  // Initialize the wallet.
  const wallet = new Wallet(privateKey);

  // Create deployer object and load the artifact of the contract you want to deploy.
  const deployer = new Deployer(hre, wallet);
  const artifact = await deployer.loadArtifact("UpgradeableGreeter");

  // Estimate contract deployment fee
  const greeting = "Hi there! This is Greeter v0";
  const deploymentFee = await deployer.estimateDeployFee(artifact, []);

  // OPTIONAL: Deposit funds to L2
  // Comment this block if you already have funds on zkSync.
  // const depositHandle = await deployer.zkWallet.deposit({
  //   to: deployer.zkWallet.address,
  //   token: utils.ETH_ADDRESS,
  //   amount: deploymentFee.mul(2),
  // });
  // Wait until the deposit is processed on zkSync
  // await depositHandle.wait();

  // Deploy this contract. The returned object will be of a `Contract` type, similarly to ones in `ethers`.
  // `greeting` is an argument for contract constructor.
  const parsedFee = ethers.utils.formatEther(deploymentFee.toString());
  console.log(`The deployment is estimated to cost ${parsedFee} ETH`);

  // Get constructor args after contract being deployed
  // import { ContractFactory, Wallet, utils } from "zksync-web3";
  // const factory = new ContractFactory(artifact.abi, artifact.bytecode, wallet);
  // const contractInterface = factory.interface;
  // const constructorArgs = getInitializerData(contractInterface, [greeting], "setGreeting");

  const greeterContract = await hreUpgrades.zkUpgrades.deployProxy(deployer.zkWallet, artifact, [greeting], { initializer: 'setGreeting' });
  await greeterContract.deployed();

  const contractInterface = greeterContract.interface;
  const constructorArgs = getInitializerData(contractInterface, [greeting], "setGreeting");
  console.log(`constructor args: ${constructorArgs}`);

  // Show the contract info.
  const contractAddress = greeterContract.address;
  console.log(`${artifact.contractName} was deployed to ${contractAddress}`);
}

### Contract Code

//SPDX-License-Identifier: Unlicense
pragma solidity ^0.8.16;
import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";

contract UpgradeableGreeter is Initializable {
    string private greeting;
    
    function initialize(string memory _init_greeting) public initializer {
        greeting = _init_greeting;
    }
    
    function greet() public view returns (string memory) {
        return greeting;
    }

    function setGreeting(string memory _greeting) public {
        greeting = _greeting;
        emit GreetingChanged(_greeting);
    }

    event GreetingChanged(string greeting);
}

### Does this work on other EVMs? (If yes, please list at least 1 of them)

yes

### Description of What Your Contract Does

Standard Greeter example, but upgradeable

### Additional Details

In the deploy script, there are 2 things I am confused about (following [this guide](https://era.zksync.io/docs/tools/hardhat/hardhat-zksync-upgradable.html#transparent-upgradable-proxies))
1. `const deploymentFee = await deployer.estimateDeployFee(artifact, []);` <- What is a way to estimate deploy fee for upgradeable? Is there such a method?
2. `hreUpgrades.zkUpgrades.deployProxy(deployer.zkWallet, artifact, [greeting], { initializer: 'setGreeting' });`. It seems like the initializer function is not called correct? This has to be called once deployed? Or did I not set it up properly for it to be initialized?
I checked the greeting string after deploying and it was empty :thin

**Comment by FilipTxFusion** [comment, marked as answer]
Hi, thanks for submitting this.

Regarding your questions:

1. We currently do not support the fee estimation for all contracts during the proxy deployment. Would you like to see the fee estimation for each separate contract deployed (implementation, proxy, admin/beacon), or would you prefer to have a total estimate of the whole deployment process? Please tell us what you would like to see here.

2. I tried replicating your issue, but it seems that all works fine. There is a `Hi there! This is Greeter v0` string when calling `greet()` method. Please make sure you call this method through a proxy contract. This is one way of how you can do it in your script:
  ``` typescript
  
  greeterContract.connect(wallet);
  const value = await greeterContract.greet();
  console.info('Greeter value is: ', value);
  
  ````

Let us know if we can help any further!

**Reply by damien75** [reply]
Thanks for the answer :)
1. I think having the breakdown of the fee might make more sense if doable (people can then sum it up if wanted) 
2. I guess I was calling the implementation contract directly and not the proxy contract :)

**Comment by newtondav** [comment]
// Answers to your questions:
// 1. The method `estimateDeployFee` in the deployer object estimates the deployment fee for an upgradeable contract.
    You can use it as shown in your code snippet: `const deploymentFee = await deployer.estimateDeployFee(artifact, []);`.
 2. Regarding the initializer function not being called correctly, ensure that the initializer function is correctly specified in the proxy deployment options.
    It seems like you've correctly specified `{ initializer: 'setGreeting' }`, but ensure that the `setGreeting` function is indeed the correct initializer for your contract.
    Also, ensure that the `initialize` function in your contract is properly defined and sets the initial state as expected.
    If the greeting string is empty after deployment, it's likely that the `initialize` function did not execute correctly.

**Comment by fullertasha** [comment]
The estimated deploy fee for upgradeable contracts can be obtained using:
const deploymentFee = await deployer.estimateDeployUpgradeFee(artifact, []);

 The initializer function is not called correctly. It should be:
const greeterContract = await hreUpgrades.zkUpgrades.deployProxy(artifact, [greeting], { initializer: 'initialize' });
